src/proxy/http/fallback.rs

```diff
@@ -0,0 +1,336 @@
+            Making::Ready(_) => return Ok(true),
+            Making::Done => panic!("polled after ready"),
+        };
+        match res {
+            Async::Ready(res) => {
+                *self = Making::Ready(res);
+                Ok(true)
+            }
+            Async::NotReady => Ok(false),
+        }
+    }
+
+    fn take(&mut self) -> T::Item {
+        match mem::replace(self, Making::Done) {
+            Making::Ready(a) => a,
+            _ => panic!(),
+        }
+    }
+}
+
+impl<P, F, A, B, C> svc::Service<http::Request<A>> for Service<P, F, A>
+where
+    P: svc::Service<http::Request<A>, Response = http::Response<B>, Error = Error<A>>,
+    F: svc::Service<http::Request<A>, Response = http::Response<C>> + Clone,
+    F::Error: Into<proxy::Error>,
+{
+    type Response = http::Response<Body<B, C>>;
+    type Error = proxy::Error;
+    type Future = ResponseFuture<P::Future, F, A>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        // We are ready iff both the primary *and* the fallback are ready.
+        // However, we always want to poll both, in case they have internal
+        // state that's driven in `poll_ready`, before returning the result.
+        let primary_ready = self.primary.poll_ready().map_err(|e| e.error);
+        let fallback_ready = self.fallback.poll_ready().map_err(Into::into);
+        try_ready!(fallback_ready);
+        primary_ready
+    }
+
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
+        ResponseFuture {
+            fallback: self.fallback.clone(),
+            state: State::Primary(self.primary.call(req)),
+        }
+    }
+}
+
+impl<P, F, A, B, C> Future for ResponseFuture<P, F, A>
+where
+    P: Future<Item = http::Response<B>, Error = Error<A>>,
+    F: svc::Service<http::Request<A>, Response = http::Response<C>>,
+    F::Error: Into<proxy::Error>,
+{
+    type Item = http::Response<Body<B, C>>;
+    type Error = proxy::Error;
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            self.state = match self.state {
+                State::Primary(ref mut f) => match f.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(rsp)) => return Ok(Async::Ready(rsp.map(Body::A))),
+                    Err(Error {
+                        fallback: Some(req),
+                        error,
+                    }) => {
+                        trace!("{}; trying to fall back", error);
+                        State::Fallback(self.fallback.call(req))
+                    }
+                    Err(e) => return Err(e.into()),
+                },
+                State::Fallback(ref mut f) => {
+                    return f
+                        .poll()
+                        .map(|p| p.map(|rsp| rsp.map(Body::B)))
+                        .map_err(Into::into)
+                }
+            }
+        }
+    }
+}
+
+impl<A, B> Payload for Body<A, B>
```

Comment:
`// === impl Body ===` would be helpful above here, or we should split into separate files